### PR TITLE
Reimplement labels with ajax option loading and label creation

### PR DESCRIPTION
**Pull request by Mathis-Z**
This allows users to create new labels and also changes the select option loading to use ajax in order to not always download all labels.

### CODE REVIEW

File: spec/controllers/tasks_controller_spec.rb
Comment (kkoehn) [inline spec/controllers/tasks_controller_spec.rb:286]: this would fit in a new context, but you can probably just set it in the `let` in line 243

We may want to add an example again where we check if a new `Label` is created.

**Review by kkoehn — CHANGES_REQUESTED** [review]
Again a good implementation, I marked some minor things
I still think, we should merge the other PR before, if it doesn't lead to merge major merge conflicts.

File: app/assets/javascripts/labels.coffee
Comment (Mathis-Z) [reply app/assets/javascripts/labels.coffee:30]: I struggled a bit with git because I had to rebase this branch onto the newer reimplement_labels branch and there were some conflicts. That is probably where this comes from.

File: app/assets/javascripts/labels.coffee
Comment (Mathis-Z) [reply app/assets/javascripts/labels.coffee:7]: I missed that. Readded it.

File: app/assets/javascripts/labels.coffee
Comment (Mathis-Z) [reply app/assets/javascripts/labels.coffee:19]: The problem with having space as a tokenSeparator is that you cannot create labels like "Data Structures" anymore because it will separate on the space. I added the colon as a separator though.

File: app/controllers/labels_controller.rb
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb:11]: select2 preloads more than one page so in order to notice the pagination you need more labels. On my system it preloads 3 pages so you would need to have more than 9 labels in your system to notice the pagination.

File: spec/controllers/tasks_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/tasks_controller_spec.rb:286]: I am not quite sure what you mean by that. Maybe you could make a code suggestion?

**Comment by Mathis-Z** [comment]
While testing the tokenSeparators for the select2 I noticed it was possible to create the same label multiple times.

File: spec/controllers/tasks_controller_spec.rb
Comment (kkoehn) [inline spec/controllers/tasks_controller_spec.rb:289]: since you already checked that the update works in the example before I would just add the following
```suggestion
      it 'creates new task label when requested' do
        expect { put_update }.to change(Label, :count).by(1)
      end
```

File: spec/controllers/tasks_controller_spec.rb
Comment (kkoehn) [reply spec/controllers/tasks_controller_spec.rb:286]: see other comment

File: spec/controllers/tasks_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/tasks_controller_spec.rb]: I moved the test into a new context. Maybe that is better.

File: spec/controllers/tasks_controller_spec.rb
Comment (kkoehn) [inline spec/controllers/tasks_controller_spec.rb:290]: Almost, but you don't really need to redefine the put:

```suggestion
        let(:changed_attributes) { {label_names: [not_existing_label_name]} }
```

Comment (kkoehn) [reply spec/controllers/tasks_controller_spec.rb:290]: I would also add the following expectation, since we identified the issue of not being exactly sure what Label gets created:

```ruby
        it 'creates a label with the correct name' do
          put_update
          expect(Label.last.name).to eql not_existing_label_name
        end
```

Comment (Mathis-Z) [reply spec/controllers/tasks_controller_spec.rb]: I committed your suggestions. Thanks for the feedback!

File: app/controllers/labels_controller.rb
Comment (MrSerth) [inline app/controllers/labels_controller.rb:13]: `results.length` is not optimal here, it will render the pagination somewhat useless. In order to understand why, one has to look what happens here:

The `length` method will get the total number of entries of an array. At this point, `results` is not an array, but rather an `ActiveRecord_Relation`. This is a wrapper for the database call, and the actual results or only lazy loaded, when needed. Since the pagination only uses the database query and further customizes it (by limiting the result size and specifying an offset), this object (and thus all label objects) still aren't loaded. However, keep in mind that the original database query is unchanged (i.e., with the limit clause). Now, you are getting the length of the result array. Hence, the original query has to be fulfilled, in order to answer this question. Consequently, all matching labels will be fetched from the database, instantiated in an object and put into this array. Once everything is done, you get the length. As you see, we were still doing the regular processing for all matching labels, and the pagination didn't kick in. In this case, it would have been better to query the database just for the one number you're interested in: The result size. This would be done with a `COUNT(*)` query, and Active Record allows you to do so easily: `results.count`. This way, we omit transferring all labels and instantiating them, and leaving the hard work to the database.

Try the difference locally, you'll notice a change in the database query:

```
Label Load (2.8ms)  SELECT "labels".* FROM "labels"
Label Count (1.6ms)  SELECT COUNT(*) FROM "labels"
```

This change is better, but still not perfect. If the `paginate` already knows how many objects we have, we aren't using the result. Hence, we should leave it to paginate to decide whether another database query is needed for the overall number of results. Do so with `paginated.total_entries`. [Here](https://github.com/mislav/will_paginate/blob/8aa3a87022e4c17cb6ebf2ec0c9467bcdfd76b18/lib/will_paginate/active_record.rb#L68-L79) is the definition of the method. As you see, it will try calculating the overall size without further querying the database if possible, and otherwise switch to a database query using a raw `COUNT`. Therefore, we should use this method.

Comment (MrSerth) [inline app/controllers/labels_controller.rb:13]: After my previous two suggestion, this line reads as:

```rb
      pagination: {more: (@page * paginated.limit_value < paginated.total_entries)},
```

I feel, this is better, but still not perfect. When checking the library further, I saw the following possibilities:

```rb
      pagination: {more: paginated.next_page.present?},
```

[According to the source](https://github.com/mislav/will_paginate/blob/8113fddf4b506867e9af3aa59fe2df314b79860c/lib/will_paginate/collection.rb#L24-L27), the `next_page` will either return the number of the next page or nil. Both is fine, and we can convert this result to the desired boolean value with `present?`.

Overall, I would go with this solution. I hope you agree that this solution is more elegant and also easier to read :).

File: app/controllers/labels_controller.rb
Comment (MrSerth) [inline app/controllers/labels_controller.rb:17]: Is there a specific reason you introduced this method? Sure, it will catch the case where no page number or a negative one is given, but is this likely to occur? There is one case you aren't checking so far: Someone could also get a page number that is larger than the total number of available pages. For the latter, WillPaginate already comes with a [helper method](https://github.com/mislav/will_paginate/blob/8113fddf4b506867e9af3aa59fe2df314b79860c/lib/will_paginate/collection.rb#L29-L35).

Since we no longer need the `@page` variable twice, I would suggest to just use the return value of this method (and without it being a before filter) if we want to keep it. Then, we should probably also add a test case 😉.

File: app/models/task.rb
Comment (MrSerth) [inline app/models/task.rb:38]: Here, we are creating new labels indirectly. We should probably also consider what happens if a label is no longer referred by any task. Currently, we are still keeping it. Is this desired ("yes" is a valid response) or should we clean unused labels?

File: spec/controllers/labels_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/labels_controller_spec.rb:15]: Rather than checking for the HTTP response no one should ever read, I would check for the actual redirect. What do you think of these expectations:

```rb
expect(response).to have_http_status(:redirect)
expect(response).to redirect_to(root_url)
```

File: spec/controllers/labels_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/labels_controller_spec.rb:20]: Usually, we used the name `get_request` for similar test cases.

Comment (MrSerth) [inline spec/controllers/labels_controller_spec.rb:20]: If you move this subject to the above `describe` context, you should be able to reuse it in the `without being signed in` context. It should work there as well, I assume 🤔

File: spec/controllers/labels_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/labels_controller_spec.rb:26]: While the test suggests that you are checking for a valid JSON, this is not exactly true. You are checking that the response matches a given pattern, which will match a valid JSON, but there are also constellations, where an invalid JSON would still match. "Just" matching for a valid JSON would be performed by trying to parse the JSON and checking for any parse errors. However, I doubt this test would be useful, since generating and parsing the JSON is mostly performed by libraries we haven't written ourselves (and thus the test would cover the library functionality...). Therefore, ask yourself: What do you _actually_ want to test here? I think you had a good intention, but just need to rework the test description `returns valid json` and the expectations.

For the expectations, the following might come handy: `response.parsed_body` will give you a hash representation of the JSON, allowing you to specify more fine granular tests. For example, you could check for some predefined keys to be present in the response or some data types (`results` should be an array, ...)

File: spec/controllers/labels_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/labels_controller_spec.rb:31]: Depending on how you proceed with the tests, this expectation could also be reworked to be more specific / the selector be more strict (rather tan just checking the whole `body`).

File: spec/controllers/tasks_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/tasks_controller_spec.rb:160]: The `eq` matcher will care about the order of elements in this array. If you ensure a stable sorting: Keep it. If you don't, you might also switch to another array matcher as outlined [here](https://medium.com/beambenefits/array-matchers-in-rspec-42b370e3983e).

File: spec/controllers/tasks_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/tasks_controller_spec.rb:285]:
```suggestion
      context 'when requesting a new label to be created' do
```

File: spec/controllers/tasks_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/tasks_controller_spec.rb:295]: I am still not a fan of the `Label.last` usage in tests. What if we search for the label by its name and expect that only one result is returned? Something like:

```
          expect(Label.where(name: not_existing_label_name).size).to be 1
```

File: spec/controllers/tasks_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/tasks_controller_spec.rb:290]: We should also add a test that reuses an existing label and ensure that the overall count isn't changed.

**Review by MrSerth — COMMENTED** [review]
Well done! Here are a few more comments, but noting major. During my review, I focussed on the new parts not present in your previous PR #898. 

Probably, it would be helpful to rebase this branch on the other one and target your previous branch, so that the diff is easier to review. Since this rebasing is not easy, I would suggest doing it together; until then, you might just focus on my comments.

File: app/assets/javascripts/labels.coffee
Comment (Mathis-Z) [reply app/assets/javascripts/labels.coffee:41]: good idea, I will do that.

File: app/assets/javascripts/labels.coffee
Comment (Mathis-Z) [reply app/assets/javascripts/labels.coffee:55]: I will extract it into a variable and also set the select2 maximumInputLength to this value to give the user a better error message.

File: app/controllers/labels_controller.rb
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb:13]: Yes, that is true.

Comment (Mathis-Z) [reply app/controllers/labels_controller.rb:13]: Just saw your other comment further down so I will change it to paginated.next_page.present? now. :)

File: app/controllers/labels_controller.rb
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb:17]: We could probably just remove the function and have rails give out an error if no page parameter is given. I did not feel comfortable have a missing parameter lead to an error but that is probably okay. 
Maybe it is even better not having the function because an error is better than silently correcting something without the caller knowing.

File: app/models/task.rb
Comment (Mathis-Z) [reply app/models/task.rb:38]: Since users can always create new labels when needed it is probably best to delete the unused ones.

File: spec/controllers/labels_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/labels_controller_spec.rb:15]: I was not aware of the redirect_to() expectation but that is much cleaner of course.

File: spec/controllers/labels_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/labels_controller_spec.rb:20]: Yes, that works.

File: spec/controllers/labels_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/labels_controller_spec.rb:26]: I changed the expectations to check for keys and variable types.

File: spec/controllers/labels_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/labels_controller_spec.rb:31]: I changed that.

File: spec/controllers/tasks_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/tasks_controller_spec.rb:160]: i switched to contain_exactly because the order is not ensured to stay the same.

File: spec/controllers/tasks_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/tasks_controller_spec.rb:295]: Thats probably better.

File: spec/controllers/tasks_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/tasks_controller_spec.rb:290]: I added a test for that.

File: app/controllers/labels_controller.rb
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb]: I just went ahead and removed the function but I am also happy to re-add it if you want.

Comment (MrSerth) [reply app/controllers/labels_controller.rb]: I think, it's fine the way it is. If no page parameter is given, WillPaginate will automatically assume page 1, similar to the method you previously added.

File: app/models/task.rb
Comment (MrSerth) [reply app/models/task.rb:38]: Yes, this probably would be desirable. The only downside I can think of is a change in the color -- since every new label gets a new random color.

File: app/assets/javascripts/labels.coffee
Comment (MrSerth) [reply app/assets/javascripts/labels.coffee]: Ah, I see; got that -- thanks. What a pity that select2 doesn't parse the data attributes of the `<option>` nodes. Then, I am fine with going for this option. Just a final thought on this topic: If we switch the order, we might improve performance: Initializing jQuery for a given element is more "expensive" than accessing an attribute of an object. Thus, if we first check for `data.label_color`, and only in the second step for `$(data.element).attr("label_color")`, we can slightly reduce the overhead of JavaScript operations (and it's not harmful anyway, I think).

File: app/models/task.rb
Comment (kkoehn) [reply app/models/task.rb:38]: We could try to generate the color from the string. Either something like
`Digest::MD5.hexdigest(label)[0..5]`


alternativly we use a hash to generate a reproduceable number and pick a color from the COLORS-array in the label-model (that way we would have more control of the available colors)
`Digest::MD5.hexdigest(label).to_i(16) % 50`

File: spec/controllers/labels_controller_spec.rb
Comment (kkoehn) [inline spec/controllers/labels_controller_spec.rb:39]: I would prefer a spec that gives an easy to read example of how the resulting json would look:

```suggestion
      it 'returns valid json' do
        get_request

        expect(JSON.parse(response.body, symbolize_names: true)).to eql(
          {
            pagination: {more: false},
            results: [{id: label.name, label_color: label.color, label_font_color: label.font_color, text: label.name}],
          }
        )
     end
```

**Review by kkoehn — COMMENTED** [review]
Great progress - I found one more point in the specs

File: app/models/task.rb
Comment (Mathis-Z) [reply app/models/task.rb:38]: I went with the second solution because it sticks to the pre-defined colors.

File: spec/controllers/labels_controller_spec.rb
Comment (Mathis-Z) [reply spec/controllers/labels_controller_spec.rb]: Thanks for the suggestion. I switched to your solution.

**Comment by MrSerth** [comment]
Ah, thanks for noticing this behavior!

Indeed, using an `after_initialize` filter would always overwrite the color persisted. Therefore, I would rather go for the following:

```rb
  validates :color, presence: true, format: { with: /\A[a-fA-F0-9]{6}\z/ }

  before_validation :choose_label_color, if: -> { color.blank? }
```

This adds a validator to ensure we always have a color and a conditional hook to add a color if not present. Still, it doesn't fix the underlying issue (but "just" prevents we are returning a 500 server error). In order to show the newly added (but not yet saved) label correctly, you should change

```rb
@task.labels.pluck(:name))
```

to 

```rb
@task.labels.map(&:name))
```

in `app/views/tasks/_form.html.slim`. The difference is that `pluck` only gets the name of persisted objects, which is not the case for the newly created label. Let's give this a try and also add a test ;).

File: spec/controllers/tasks_controller_spec.rb
Comment (MrSerth) [inline spec/controllers/tasks_controller_spec.rb:241]: In this test, we should also check that the `not_existing_label_name` is present and *checked* (Hint: Write a test that fails without the `@task.labels.map(&:name))` change, but only succeeds otherwise. If you test this manually, you'll quickly notice the difference in the HTML -- that is the `selected` attribute for this `option`).

Comment (Mathis-Z) [reply spec/controllers/tasks_controller_spec.rb:241]: I think that would be a good integration test. Something like this:
```
expect(rendered).to have_selector('option[value="some new label"][selected="selected"]')
```

Other than this I found rails-dom-testing to be in the Gemfile but could not find any usages in CodeHarbor and also no examples of how to use it online.
The last option I could think of is using regex again although that is probably not a good idea.

Comment (MrSerth) [reply spec/controllers/tasks_controller_spec.rb:241]: Indeed, I would also go for an integration test. Let's use this as an example and follow up on it next week ;).

The `rails-dom-testing` is only included as an indirect dependency in the `Gemfile.lock`, and therefore not (directly) used so far.

**Review by MrSerth — APPROVED** [review]
Awesome, thanks for taking care of the test as well and adding these nice features! 🎷

**Pull request merged**
